### ISSUE TITLE
LSP references results deduplication

### ISSUE THREAD

**Issue by joeveiga** [nvim-telescope/telescope.nvim#3328]
**Is your feature request related to a problem? Please describe.**
After #3211, now I'm seeing duplicate results in my lsp references results:

![image](https://github.com/user-attachments/assets/c1343079-053c-4405-8e70-b1b6f97dd2ec)

It looks like this was a known issue when the feature was added. From the PR description:
> This leaves us with the issue of what happens if both LSPs returning overlapping locations, so deduplication could be in order
But since this isn't my case and I suspect not a lot of people use multiple LSPs that return the same stuff, I prefer to get your feedback first and maybe implement it later

Unfortunately in my usecase I do need a way to dedup the list :). I'm using both `angularls` and `ts_ls`, and they often return the same locations.
 
**Describe the solution you'd like**
It'd be useful to have an option to remove duplicates from the results (probably enabled by default, even).

**Comment by jamestrew** [comment]
Is angularls and ts_ls intended to be used at the same time on the same buffers?
Seems strange to me that two language server would yield identical references.

Trying to think what the best way of handling this is. Curious how `:lua vim.lsp.buf.references()` handles this.

**Comment by joeveiga** [comment]
This is the output from `vim.lsp.buf.references()` on the same symbol as the previous screenshot with telescope:

![image](https://github.com/user-attachments/assets/2d485570-d1c3-433d-8a8e-2082eede64a1)

Looks like there aren't any duplicates.

**Comment by jamestrew** [comment]
If you do `:chistory` do you see two quickfix lists for lsp reference?

**Comment by joeveiga** [comment]
Hey @jamestrew. Yeah, I get two:

![image](https://github.com/user-attachments/assets/08f40584-5f1b-4220-9144-bb6768143d79)

**Comment by jamestrew** [comment]
Looks like neovim is merging references (and generally handling multiple clients for various lsp handlers similar to how telescope has started doing this) now.
https://github.com/neovim/neovim/pull/30722

It doesn't look like there's any de-duping logic there. I'm curious if they will add it and if so what the implementation will be.

**Comment by at-karan-sharma** [comment]
I also had the same duplication issue with `angularls` and `ts_ls`. Thanks to the discussion here, I tried updating neovim to the latest pre-release build and it fixed the problem.

**Comment by at-karan-sharma** [comment]
Looks like this issue is still present on the latest neovim build, even after neovim/neovim#30722.

Here's a workaround:
```
local function filterDuplicates(array)
    local uniqueArray = {}
    for _, tableA in ipairs(array) do
        local isDuplicate = false
        for _, tableB in ipairs(uniqueArray) do
            if vim.deep_equal(tableA, tableB) then
                isDuplicate = true
                break
            end
        end
        if not isDuplicate then
            table.insert(uniqueArray, tableA)
        end
    end
    return uniqueArray
end

local function on_list(options)
    options.items = filterDuplicates(options.items)
    vim.fn.setqflist({}, ' ', options)
    vim.cmd('botright copen')
end

# Usage
vim.lsp.buf.references(nil, { on_list = on_list })
```

Bonus: a related issue with this scenario is `vim.lsp.buf.rename()` asks twice when renaming a variable in a .ts file. This can be fixed by specifying the desired client via `vim.lsp.buf.rename(nil, { name = 'angularls' })`

**Comment by davidecavaliere** [comment]
This is my solution:

```lua
local function is_angular(filename)

  return lspconfig.util.root_pattern("angular.json")(filename);
end

local function is_deno(filename)

  return lspconfig.util.root_pattern("deno.json", "deno.jsonc")(filename);
end

lspconfig.ts_ls.setup({
  on_attach = function (client, bufnr)
    nvlsp.on_attach(client, bufnr);
    vim.keymap.set('n', 'gro', organizeImports, { buffer = bufnr,  remap = false, desc = "Reorganize imports"});
  end,
  root_dir = function (filename, bufnr)
    if is_deno(filename) or is_angular(filename) then
      -- if it's deno we want to let deno_ls handle it
      -- angular-ls already implement ts_ls: letting them both handle the file
      -- will result in duplicated entries in references or duplicated calls to any code action
      return nil;
    end

    return lspconfig.util.root_pattern("package.json")(filename);
  end,
  on_new_config = function(new_config,new_root_dir)
  
  single_file_support = false,
  config = function (options)
    vim.print(options);
    return options;
  end
})

```

Edit: 
This is actually not a viable solution for angular_ls. Because while some features from the angular_ls duplicates the ones from ts_ls there is not a complete overlap so excluding the one or the other will lead to missing features.

It is still a valid solution to avoid conflicts between ts_ls and deno_ls

**Comment by alder-french-leviton** [comment]
Currently experiencing this problem with a Neovim kickstart setup when using angularls and ts_ls simultaneously. Since I'm using Mason and the other kickstart lsp defaults, I'm not able to get the deduplication workaround by @at-karan-sharma working when I add it to my init.lua file.

**Comment by french-fryy** [comment]
Currently experiencing this problem with a Neovim kickstart setup when using angularls and ts_ls simultaneously. Since I'm using Mason and the other kickstart lsp defaults, I'm not able to get the deduplication workaround by @at-karan-sharma working when I add it to my init.lua file. Instead I updated the "lua/telescope/builtin/\_\_lsp.lua" with his filterDuplicates() function and it worked. Obviously this isn't a permanent solution. I'm thinking what might be best is adding an additional option like "entry_filter_function" to "lsp.references" and "list_or_jump" that allows a user to pass a filter function for handling things like duplicate references, etc...
Here's a code snippet that shows my modified code in "lua/telescope/builtin/\_\_lsp.lua":
```lua
--NOTE: filterDuplicates below added and called in list_or_jump to filter duplicates for goto references -Alder
local function filterDuplicates(array)
  local uniqueArray = {}
  for _, tableA in ipairs(array) do
    local isDuplicate = false
    for _, tableB in ipairs(uniqueArray) do
      if vim.deep_equal(tableA, tableB) then
        isDuplicate = true
        break
      end
    end
    if not isDuplicate then
      table.insert(uniqueArray, tableA)
    end
  end
  return uniqueArray
end

---@param action telescope.lsp.list_or_jump_action
---@param title string prompt title
---@param funname string: name of the calling function
---@param params lsp.TextDocumentPositionParams|(fun(client: vim.lsp.Client, bufnr: integer): table?)
---@param opts table
local function list_or_jump(action, title, funname, params, opts)
  opts.reuse_win = vim.F.if_nil(opts.reuse_win, false)
  opts.curr_filepath = vim.api.nvim_buf_get_name(opts.bufnr)

  vim.lsp.buf_request_all(opts.bufnr, action, params, function(results_per_client)
    local items = {}
    local first_encoding
    local errors = {}

    for client_id, result_or_error in pairs(results_per_client) do
      local error, result = result_or_error.error, result_or_error.result
      if error then
        errors[client_id] = error
      else
        if result ~= nil then
          local locations = {}

          if not utils.islist(result) then
            vim.list_extend(locations, { result })
          else
            vim.list_extend(locations, result)
          end

          local offset_encoding = vim.lsp.get_client_by_id(client_id).offset_encoding

          if not vim.tbl_isempty(result) then
            first_encoding = offset_encoding
          end

          vim.list_extend(items, vim.lsp.util.locations_to_items(locations, offset_encoding))
        end
      end
    end

    for _, error in pairs(errors) do
      vim.api.nvim_err_writeln("Error when executing " .. action .. " : " .. error.message)
    end

    items = apply_action_handler(action, items, opts)
    items = filter_file_ignore_patters(items, opts)
    items = filterDuplicates(items) --NOTE: this line added by Alder

    if vim.tbl_isempty(items) then
    ...


```

**Comment by RobinMeow** [comment]
im also having this issue using angular_ls with tsc
angular_ls doesnt not implement all features which tsc implements, to disabling one is not a solution.
The deduplicate thing would be good enough for me, but would be nice to have a `enabled` flag, becuase I couldnt get it working. Based on this on all the others issue spread acrros neovim and telescope repo.
I know the issue is **me** that I couldnt get it working. Because I dont know neovim well enoug yet.

I tried using the solution from the [docs](https://neovim.io/doc/user/lsp/):
```lua
  local function on_list(what)
    vim.list.unique(what.items, function(item)
      return ("%s\0%d\0%d\0%d\0%d"):format(
        item.filename or "",
        item.lnum or 0,
        item.col or 0,
        item.end_lnum or 0,
        item.end_col or 0
      )
    end)

    vim.fn.setqflist({}, " ", what)
    vim.cmd("botright copen")
  end
  vim.lsp.buf.definition({ on_list = on_list })
  vim.lsp.buf.references(nil, { on_list = on_list })
```

`vim.cmd("botright copen")` obv shouldnt be used with telescope, otherwise you have the telescope picker and the qflist open.

and into on_attach doesnt work either. (As I said, a **me** issue)
I think I have to set it into everykeymap, which sounds like a big annoyance to me
` map("grr", require("telescope.builtin").lsp_references, "[G]oto [R]eferences")`
how do I put this in here, without having to re-implement lsp_references?

---

I did get the double rename prompt workaround woking from @at-karan-sharma 
```lua
local map = function(keys, func, desc, mode)
  mode = mode or "n"
  vim.keymap.set(mode, keys, func, { buffer = event.buf, desc = "LSP: " .. desc })
end

-- from this
map("grn", vim.lsp.buf.rename, "[R]e[n]ame")
-- to this
map("grn", function()
  vim.lsp.buf.rename(nil, { name = "angularls" })
end, "[R]e[n]ame")
```
tho I think without checking for which lsp is attached, this will probably screw up my non-angular project renames

---

having a picker for which lsp to use for actions like rename, would be amazing (set one by default ,and use a seperate keybind for choosing an lsp before doing the rename)

this is ai generated code which allows me to pick a ls, for rename using telescope picker. works for me
```lua
          local map = function(keys, func, desc, mode)
            mode = mode or "n"
            vim.keymap.set(mode, keys, func, { buffer = event.buf, desc = "LSP: " .. desc })
          end
          map("grn", function()
            local clients = vim.tbl_filter(function(c)
              return c:supports_method("textDocument/rename")
            end, vim.lsp.get_clients({ bufnr = 0 }))

            require("telescope.pickers")
              .new({}, {
                prompt_title = "LSP Rename",
                finder = require("telescope.finders").new_table({
                  results = clients,
                  entry_maker = function(c)
                    return { value = c, display = c.name, ordinal = c.name }
                  end,
                }),
                sorter = require("telescope.config").values.generic_sorter({}),
                attach_mappings = function(_, map)
                  local actions = require("telescope.actions")
                  local state = require("telescope.actions.state")

                  map("i", "<CR>", function()
                    local c = state.get_selected_entry().value
                    actions.close(_)
                    vim.lsp.buf.rename(nil, { name = c.name })
                  end)
                  return true
                end,
              })
              :find()
          end, "Rename")
```
